### PR TITLE
PICARD-1601: Ensure we package for PyPI with all test dependencies

### PR DESCRIPTION
**Pull request by phw**
Ensure we package for PyPI with all test dependencies.

# Summary
Fully test packaging sdist, extracting it and running the tests inside it. Fixes that we miss packaging test dependencies.

* This is a…
    * [x] Bug fix
    * [ ] Feature addition
    * [ ] Refactoring
    * [ ] Minor / simple change (like a typo)
    * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

* JIRA ticket (_optional_): [PICARD-1601](https://tickets.metabrainz.org/browse/PICARD-1601)

# Solution

# Action

### CODE REVIEW

**Comment by phw** [comment]
This is just a first commit to see the tests failing. I have another commit waiting which actually ads missing files to the manifest.

**Comment by phw** [comment]
I have no idea why tests are failing. Locally this is running just fine, but on Travis extracting the archive and running setup.py test (or py.test) on it just fails with module not found issues like:

```
___________________ ERROR collecting test/test_filesystem.py ___________________
ImportError while importing test module '/home/travis/build/metabrainz/picard/dist/picard-2.2.0.dev4/test/test_filesystem.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
test/test_filesystem.py:7: in <module>
    from test.picardtestcase import PicardTestCase
E   ModuleNotFoundError: No module named 'test.picardtestcase'
```

I don't get it. This is in the correct directory, the files in test are there. I don't see the difference to my local setup.

**Comment by zas** [comment]
> I have no idea why tests are failing. Locally this is running just fine, but on Travis extracting the archive and running setup.py test (or py.test) on it just fails with module not found issues like:
> 
> ```
> ___________________ ERROR collecting test/test_filesystem.py ___________________
> ImportError while importing test module '/home/travis/build/metabrainz/picard/dist/picard-2.2.0.dev4/test/test_filesystem.py'.
> Hint: make sure your test modules/packages have valid Python names.
> Traceback:
> test/test_filesystem.py:7: in <module>
>     from test.picardtestcase import PicardTestCase
> E   ModuleNotFoundError: No module named 'test.picardtestcase'
> ```
> 
> I don't get it. This is in the correct directory, the files in test are there. I don't see the difference to my local setup.

I don't see `test/__init__.py` in copied files, not sure if it is intended.

**Comment by phw** [comment]
> I don't see `test/__init__.py` in copied files, not sure if it is intended.

Good catch. It is here locally, but missing on the travis package. Confusing, but I'll look into it.

**Comment by phw** [comment]
@zas Found the issue. I accidentally changed:

    recursive-include test *.py

to 

    recursive-include test/data *.py

Not sure why this works locally, I tried several times with completely removing the dist dir. But it is definitely what broke it on travis. Thanks for finding this, it is fixed now.

**Comment by samj1912** [comment]
I would suggest putting this on as an extra target. https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

**Comment by phw** [comment]
We had issues reported for the source distribution not being able to run the tests, and I think running the tests before installing something from source is a valid use case. Would the extras solve this?

File: scripts/run-sdist-test.sh [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+# Build the sdist archive, extract it and run the tests.
+
+set -e
+
+rm -rf dist
+python3 setup.py sdist
+cd dist
+SDIST_ARCHIVE=$(ls picard-*.tar.gz)
```
Comment (zas) [inline scripts/run-sdist-test.sh]: This use of `ls` looks suspicious to me. Using `echo` builtin is prolly safer.

**Comment by samj1912** [comment]
@phw I think this belongs in extras. I have never seen a pip package that includes tests. If they do, it's always as an extra target.

**Comment by samj1912** [comment]
The python package authority also recommends excluding tests in their sample project. https://github.com/pypa/sampleproject/blob/master/setup.py#L133

**Comment by samj1912** [comment]
And they have an extra target here to install test dependencies https://github.com/pypa/sampleproject/blob/master/setup.py#L160

**Comment by phw** [comment]
@samj1912 Ok, might make sense then. Can you add a ticket for this? My main concern right now is fixing the existing package (which includes tests) since we are about to make a release, we can optimize it later.

On a quick test on my system tests are always included, no matter if they are set in the manifest or not. Probably also the reason why the same build succeeded for me while failing on Travis. 

Also where we really could look  into to not include everything is resources. Actually most of the images there are not needed in the sdist as they are build into the Python code.

**Comment by samj1912** [comment]
https://tickets.metabrainz.org/browse/PICARD-1602

**Comment by phw** [comment]
Thanks a lot @samj1912 

See also my notes on https://tickets.metabrainz.org/browse/PICARD-1602 , things are a bit strange. @mineo  what are your ideas here?

For now I'm going to merge this PR so we at least have a working sdist, even if it contains more than necessary.

**Pull request merged**